### PR TITLE
ASoC: SOF: ipc: introduce cont_update_posn in sof_ipc_stream_params s…

### PR DESCRIPTION
**Pull request by yaochunhung**
…truct

The host stream position is updated when no_stream_position is set as 0.
However current implementation updates host stream position only when report
data is larger than or equal to host period size which is decided by the
period size of host side. It maybe cause host stream position update not
in time. Therefore this patch introduces another field "cont_update_posn",
a boolean value aimed to update host stream position continuously and based
on period size of pipeline. It can get better precise when need to update
host stream position from firmware.

Signed-off-by: YC Hung <yc.hung@mediatek.com>

### CODE REVIEW

**Comment by yaochunhung** [comment]
Please reference the discussion in https://github.com/thesofproject/sof/pull/5137#discussion_r777324226  Thank.

File: include/sound/sof/stream.h
Comment (dbaluta) [inline include/sound/sof/stream.h:89]: Could we just take only 1 byte from reserved area. We are wasting so much space to keep a boolean information.

Comment (yaochunhung) [reply include/sound/sof/stream.h:89]: @dbaluta yes, I could modify it as one byte. Thanks

**Comment by lgirdwood** [comment]
Corresponding FW PR is https://github.com/thesofproject/sof/pull/5137

**Comment by lgirdwood** [comment]
@yaochunhung ABI is 3.21, this will need updated in header file.

**Comment by yaochunhung** [comment]
@lgirdwood  The original abi in kernel looks like different from firmware(kernel only 3.18, firmware 3.20). Should I need to make them consistent? Thanks
https://github.com/thesofproject/linux/blob/topic/sof-dev/include/uapi/sound/sof/abi.h#L29

**Comment by yaochunhung** [comment]
@lgirdwood  I have updated abi version v3.21. please help to check it. thanks.

**Comment by kv2019i** [comment]
@yaochunhung wrote:
> @lgirdwood The original abi in kernel looks like different from firmware(kernel only 3.18, firmware 3.20). Should I need to make them consistent? Thanks https://github.com/thesofproject/linux/blob/topic/sof-dev/include/uapi/sound/sof/abi.h#L29

The kernel side header is currently a bit under discussion, see https://github.com/thesofproject/linux/pull/3298
and https://github.com/thesofproject/linux/pull/3296 .

~~Let's try to merge PR3298 first and then you can bump to 3.21 in this PR.~~
@yaochunhung I have up on PR3296. Please just bump the version to 3.21 in your PR directly.

File: include/sound/sof/stream.h
Comment (kv2019i) [inline include/sound/sof/stream.h:87]: @yaochunhung Isn't the description in git commit message reversed? I mean if "no_stream_position" is set as 1, FW will not send position IPC. Further, this has been deprecated in ABI3.10 (see commit "ASoC: SOF: Intel: hda: fix reset of host_period_bytes"). Or do you mean something else with "host stream position is updated" ? I'll add more comments to the FW side PR, maybe it's easier to discuss there.

File: include/sound/sof/stream.h
Comment (ujfalusi) [reply include/sound/sof/stream.h:89]: We will set the `cont_update_posn` to one in kernel unconditionally or what is the condition to set it?
Iow, will the use of this new flag going to be:
```
+ 	if (v->abi_version >= SOF_ABI_VER(3, 21, 0))
+  		msg->cont_update_posn = 1;
```
If it is then I argue that there is no point of having this flag in the first place and just do it unconditionally in firmware?

**Review by kv2019i — CHANGES_REQUESTED** [review]
Thanks @yaochunhung , this looks like a good addition to the IPC interface. I have some comments on how this is implemented, please see inline.

File: include/sound/sof/stream.h
Comment (yaochunhung) [reply include/sound/sof/stream.h:87]: @kv2019i  Thanks for comment, I think commit message should be wrong and need to be modified as The host stream position is updated when no_stream_position is set as "0".  I think the commit "ASoC: SOF: Intel: hda: fix reset of  host_period_bytes" which deprecated no_stream_position in ABI3.10 only influences intel's platforms. For mediatek's platform, it is set as default 0 and DSP fw will update host position periodically.

File: include/sound/sof/stream.h
Comment (yaochunhung) [reply include/sound/sof/stream.h:89]: @ujfalusi Currently I propose to set it in the stream callback function pcm_hw_params e.g.
```c
#define CONTINUOUS_UPDATE_POSITION 1

static int mt8195_dsp_pcm_hw_params(struct snd_sof_dev *sdev,
			  struct snd_pcm_substream *substream,
			  struct snd_pcm_hw_params *params,
			  struct sof_ipc_stream_params *ipc_params)
{
	struct sof_ipc_fw_version *v = &sdev->fw_ready.version;

	if (v->abi_version >= SOF_ABI_VER(3, 21, 0))
		ipc_params->cont_update_posn = CONTINUOUS_UPDATE_POSITION;

	return 0;
}
```

**Comment by plbossart** [comment]
@yaochunhung I am not sure I understand the ask, and this seems to be an IPC interface change without the matching firmware changes?

If the idea was to have the firmware send an IPC message, hence an interrupt, on a very frequent basis, then this idea will by construction get in the way of power savings. You might want to use the memory windows or some sort of shared SRAM to provide the precise position instead. It's my understanding that ChromeOS relies on the no-interrupt mode when available, so it's be rather odd to have an interrupt added just to add more precision to the elapsed time?

**Comment by yaochunhung** [comment]
@plbossart I only update IPC message more frequently and send interrupt to host based on original host period size. Please see firmware patch https://github.com/thesofproject/sof/pull/5137/commits/af602b11e5132e3233671e415d9015a1c01c85a2 
In host side, I implement pcm pointer callback as below(Sorry I didn't send this PR yet due to it is related with https://github.com/thesofproject/linux/pull/3258 since I need to modify mt8195_ipc_msg_data)
```c
snd_pcm_uframes_t mtk_adsp_pcm_pointer(struct snd_sof_dev *sdev,
				       struct snd_pcm_substream *substream)
{
	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
	struct snd_soc_component *scomp = sdev->component;
	struct snd_sof_pcm *spcm;
	struct sof_ipc_stream_posn posn;
	struct snd_sof_pcm_stream *stream;
	snd_pcm_uframes_t pos;
	int ret;

	spcm = snd_sof_find_spcm_dai(scomp, rtd);
	if (!spcm) {
		dev_warn_ratelimited(sdev->dev, "warn: can't find PCM with DAI ID %d\n",
				     rtd->dai_link->id);
		return 0;
	}

	stream = &spcm->stream[substream->stream];
	ret = snd_sof_ipc_msg_data(sdev, stream->substream, &posn, sizeof(posn));
	if (ret < 0) {
		dev_warn(sdev->dev, "failed to read stream position: %d\n", ret);
		return 0;
	}

	memcpy(&stream->posn, &posn, sizeof(posn));
	pos = spcm->stream[substream->stream].posn.host_posn;
	pos = bytes_to_frames(substream->runtime, pos);

	return pos;
}
````

**Comment by plbossart** [comment]
> @plbossart I only update IPC message more frequently and send interrupt to host based on original host period size. Please see firmware patch [thesofproject/sof@af602b1](https://github.com/thesofproject/sof/commit/af602b11e5132e3233671e415d9015a1c01c85a2) In host side, I implement pcm pointer callback as below(Sorry I didn't send this PR yet due to it is related with #3258 since I need to modify mt8195_ipc_msg_data)


I don't follow at all, your reply mentions both interrupts generated by the firmware and IPC initiated by the kernel. which is it?
What does 'original host period size' mean? We don't modify period sizes in SOF, we take what the application specified. Are you talking about the *firmware* host period, which represents a buffer in DSP memory, or a *kernel* period, which represents a buffer in system DDR?

if you only send an IPC from the kernel to the firmware on a .pointer callback, then what does this patch actually do?

conversely, if the firmware can send notifications to the kernel on a position update, then why would you need to request the position on a .pointer as well?

Too many questions as you can see...

**Comment by yaochunhung** [comment]
@plbossart For mt8195, I set no_stream_position as zero and DSP will send stream position in the host_update_position function. where I said "original host period size" is the host_period_bytes which is set in host_params(the value is sent from host side, sof_pcm_hw_params: pcm.params.host_period_bytes = params_period_bytes(params);)
```c
=== DSP firmware part ===
src/audio/host.c
		if (hd->host_period_bytes != 0 &&
		    hd->report_pos >= hd->host_period_bytes) {
			hd->report_pos = 0;

			/* send timestamped position to host
			 * (updates position first, by calling ops.position())
			 */
			pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
			mailbox_stream_write(dev->pipeline->posn_offset,
					     &hd->posn, sizeof(hd->posn));
			ipc_msg_send(hd->msg, &hd->posn, false);
		}
```
In the .pointer callback function, I use snd_sof_ipc_msg_data to read the host stream position.
```c
===kernel part===
sound\soc\sof\mediatek\mt8195\mt8195.c)
static int mt8195_ipc_msg_data(struct snd_sof_dev *sdev,
				struct snd_pcm_substream *substream,
				void *p, size_t sz)
{
	if (!substream || !sdev->stream_box.size) {
		sof_mailbox_read(sdev, sdev->dsp_box.offset, p, sz);
	} else {
		struct sof_mtk_adsp_stream *mstream = substream->runtime->private_data;

		/* The stream might already be closed */
		if (!mstream)
			return -ESTRPIPE;

		sof_mailbox_read(sdev, mstream->stream.posn_offset, p, sz);
	}

	return 0;
}
````
For why need to provide this callback function. It is because when I test alsa_conformance_test, I found there are two points to update hw pointer. One is snd_pcm_period_elapsed, and it is called by IPC notification from DSP(SOF_IPC_STREAM_POSITION). Another is called from snd_pcm_common_ioctl(snd_pcm_hwsync) and it is not triggered by DSP IPC notification(Maybe somewhere in the host but I am not familiar with alsa_conformane_test codes and not sure how it can trigger snd_pcm_hwsync.) Therefore I implement this .pointer callback function to get more precisive value from DSP side.  Thanks.

**Review by plbossart — CHANGES_REQUESTED** [review]
Not getting the idea, see comments on the firmware side. You have not addressed my concerns on the number of interrupts generated by the firmware.

**Comment by yaochunhung** [comment]
@plbossart  Please see my comment on the firmware side. thanks.

**Comment by cujomalainey** [comment]
@plbossart This may not be the correct solution to the problem, but I am guessing that the default solution where additional HW is not available is not enough to pass the conformance test. E.g. I grabbed a samus (BDW) running SOF which does not have the pointer callback and came back with similar results.

```
localhost ~ # alsa_conformance_test -P hw:1,0 
---------PRINT PARAMS---------
PCM name: hw:1,0
card: rt5677 [sof-bdw rt5677]
device: Low Latency (*) []
stream: PLAYBACK
merge_threshold_t: 0.000100
merge_threshold_sz: 1024
access type: MMAP_INTERLEAVED
format: S16_LE
channels: 2
rate: 48000 fps
period time: 21333 us
period size: 1024 frames
buffer time: 341333 us
buffer size: 16384 frames
---------TIMER RESULT---------
                                 Total_time(s)               Counts          Averages(s)
snd_pcm_open                       0.001123312                    2             0.000562
snd_pcm_hw_params                  0.002771299                    2             0.001386
snd_pcm_hw_params_any              0.000017010                    2             0.000009
snd_pcm_sw_params                  0.000099919                    2             0.000050
snd_pcm_prepare                    0.000045145                    2             0.000023
snd_pcm_start                      0.000201959                    2             0.000101
snd_pcm_avail                      1.866766516              2528131             0.000001
precision: 0.000000001
----------RUN RESULT----------
number of recorders: 1
number of points: 47
step average: 1024.000000
step min: 1024
step max: 1024
step median: 1024
step standard deviation: 0.000000
rate: 48006.917916
rate error: 15.654371
number of underrun: 298
number of overrun: 0
```

File: include/sound/sof/stream.h
Comment (plbossart) [reply include/sound/sof/stream.h:89]: I think this should be the default across all platforms, I don't really see a reason why this capability needs to be added 4 times when we all have to support the Chrome conformance suite. @cujomalainey any comments?

Comment (cujomalainey) [reply include/sound/sof/stream.h:89]: agreed, given we know the current default does not pass the test, lets push this into the core and replace the existing default

**Comment by plbossart** [comment]
@cujomalainey for some reason I am only seeing your comment now.

For  Intel SKL+ platforms, the pointer is given by the DMA DPIB registers, without even looking at the firmware reports. That means that absolutely no one on the Intel side has ever checked what the firmware reported. That's probably why no one is able to respond on the relationship between no_stream_position and the suggested cont_update_posn

For BYT/BDW, I am not sure I get the problem. we added the flags SNDRV_PCM_INFO_BATCH to say that the precision is not very good outside of the period_elapsed events, and this is used by PulseAudio to disable timer-based scheduling. We also did not advertise support for NO_PERIOD_WAKEUP, I am not sure what your expectations on timestamps might be in this context.

**Comment by cujomalainey** [comment]
@plbossart its just proof that the baseline implementation is not enough to meet chromespec. Nothing needs to be done for the platforms, I just think we need to be open to a implementation that works for platforms that don't have access to the DMA registers.

**Comment by plbossart** [comment]
> @plbossart its just proof that the baseline implementation is not enough to meet chromespec. Nothing needs to be done for the platforms, I just think we need to be open to a implementation that works for platforms that don't have access to the DMA registers.

@cujomalainey what exactly is the chromespec requirement? I am not able to figure out what the log above with a bunch of number means. Are you saying that an implementation which only provides an accurate position every 20ms is disqualified?

**Comment by cujomalainey** [comment]
I can't find any published spec but I know the above test is part of our testing infra so that will fail if when we bring this in. Let me sync up internally and come back.

**Comment by cujomalainey** [comment]
@chiang831 pointed me to the correct docs. It is listed in the android spec which we also adhere to

* [C-1-1] The output timestamp returned by AudioTrack.getTimestamp and AAudioStream_getTimestamp is accurate to +/- 2 ms.

From https://source.android.com/compatibility/12/android-12-cdd#56_audio_latency

**Comment by plbossart** [comment]
> @chiang831 pointed me to the correct docs. It is listed in the android spec which we also adhere to
> 
> * [C-1-1] The output timestamp returned by AudioTrack.getTimestamp and AAudioStream_getTimestamp is accurate to +/- 2 ms.
> 
> From https://source.android.com/compatibility/12/android-12-cdd#56_audio_latency

@cujomalainey The 'accuracy' is relative since it may not include the delay information. From the AudioTimeStamp documentation:

"The time is based on the implementation's best effort, using whatever knowledge is available to the system, but cannot account for any delay unknown to the implementation."

if we only use the timestamp at the host DMA level then we ignore all delays within the firmware.... If you really want accuracy all the way to the audio interfaces, SOF needs a lot more infrastructure to report internal delays.

**Comment by cujomalainey** [comment]
@plbossart delay in the FW is fine. We just need more granularity when we don't have the dma on the host side to query. The alsa_conformance_test by default will report an underrun if it can't get an update at least every 5ms.

**Comment by ujfalusi** [comment]
> @plbossart delay in the FW is fine. We just need more granularity when we don't have the dma on the host side to query. The alsa_conformance_test by default will report an underrun if it can't get an update at least every 5ms.

The problem is when your hardware FIFO and DMA burst is deeper/longer than 5ms... In that case one _must_ implement the delay as the conformance test needs something to be changed every 5ms, right?

**Comment by plbossart** [comment]
> > @plbossart delay in the FW is fine. We just need more granularity when we don't have the dma on the host side to query. The alsa_conformance_test by default will report an underrun if it can't get an update at least every 5ms.
> 
> The problem is when your hardware FIFO and DMA burst is deeper/longer than 5ms... In that case one _must_ implement the delay as the conformance test needs something to be changed every 5ms, right?

that was my understanding, e.g. for a 'deep buffer' PCM path with e.g. 100ms stored, the conformance would fail if we only looked at the host DMA updates.

**Comment by cujomalainey** [comment]
Hmm, I see your point here, but what about a low latency buffer?  How would that handle in these infrequent updates. For the record the test code is [here](https://chromium.googlesource.com/chromiumos/platform/audiotest/+/refs/heads/main/alsa_conformance_test/alsa_conformance_thread.c#312) if you wish to see what it is doing.

**Comment by ujfalusi** [comment]
Low latency != frequent update.
Frequent update is a consequence of a bypassed or small FIFO. If you use bigger FIFO you can not have low latency (constantly low latency).

If you configure the period size to be 20ms and expect at least 5ms DMA position update then it might not work on all hardware. No update within 1.5*period_time is reasonable assumption for a failure.
I remember that it was discussed in a past that ALSA should be able to report the DMA burst size for user space, that is, how much it is 'jumping' when it is servicing a FIFO.

Low latency == no/small FIFO -> small jumps in DMA to keep the FIFO or interface fed.
In low latency configuration the delay is negligible by definition, the delay became more important when the FIFO is larger and user space want to get the answer to the following question: How long it is going to take for the sample pointed by the DMA at the moment to reach the speaker.
If you have burst filled FIFO then the answer to this question between bursts (when the DMA is not moving) is decreasing with time, then after a burst it is again going to be a big number.

If only the DSP side knows the position of the DMA in the host buffer then yes, it makes a lot of sense to make this information available to the host.

File: include/sound/sof/stream.h
Comment (plbossart) [reply include/sound/sof/stream.h:89]: @yaochunhung can you please add a user for this new capability?

**Comment by bhiregoudar** [comment]
> @plbossart delay in the FW is fine. We just need more granularity when we don't have the dma on the host side to query. The alsa_conformance_test by default will report an underrun if it can't get an update at least every 5ms.



> @plbossart delay in the FW is fine. We just need more granularity when we don't have the dma on the host side to query. The alsa_conformance_test by default will report an underrun if it can't get an update at least every 5ms.

Not every 5 ms, we have observed alsa_confirmance reporting under run even for 1ms and 2 ms period/frame length also. alsa queries many times within 1 ms and expects the updated position value or the smaller latency in reporting back the position info. When IPC message is used to query the position info, alsa_conformance test fails due to latency in reporting back the position info. Continuous update of position info in stream box is helping to reduce the latency and alsa_conformance test is passing. But this will have additional performance impact as position info is written to stream box for every period/frame length.

**Comment by dbaluta** [comment]
@yaochunhung what is the status of this? I noticed that the corresponding ABI change was merged into FW.

File: include/uapi/sound/sof/abi.h
Comment (dbaluta) [inline include/uapi/sound/sof/abi.h:32]: Did we forget to bump the ABI on Linux side in the past?

**Comment by yaochunhung** [comment]
> @yaochunhung what is the status of this? I noticed that the corresponding ABI change was merged into FW.

@dbaluta  I don't have any concern about this PR but I saw there were some discussion above.
@cujomalainey @plbossart @ujfalusi @bhiregoudar , could you please check if there are any suggestions or concerns about this PR? Thanks.

**Comment by cujomalainey** [comment]
Apologies, I am wrapping up some internal projects. I think this is something we need to address at a TSC then spin off into its own design discussion given the implications it has on the IPC window.

**Comment by lgirdwood** [comment]
> > @yaochunhung what is the status of this? I noticed that the corresponding ABI change was merged into FW.
> 
> @dbaluta I don't have any concern about this PR but I saw there were some discussion above. @cujomalainey @plbossart @ujfalusi @bhiregoudar , could you please check if there are any suggestions or concerns about this PR? Thanks.

~~Sorry, confused - this is not yet merged into FW ? PR is still open. Do we mean it's merged in drivers now and NOT FW ?~~

~~Or we mean mean the big bump in kernel MINOR from 18 to 21 ?~~

EDIT, sorry lets start again, I had assumed this was FW PR and not kernel PR.

**Comment by yaochunhung** [comment]
@lgirdwood  The PR of FW have already been merged https://github.com/thesofproject/sof/commit/6a18e5a8c95d4583162ad42cadc3585ad9434b3d. Currently this PR for Linux is not merged yet. Thanks.

**Comment by lgirdwood** [comment]
> @lgirdwood The PR of FW have already been merged [thesofproject/sof@6a18e5a](https://github.com/thesofproject/sof/commit/6a18e5a8c95d4583162ad42cadc3585ad9434b3d). Currently this PR for Linux is not merged yet. Thanks.

Thanks, I got the target mixed. 

So this looks fine for uses cases or platforms where we need a consistent position available to host via either IPC, mailbox or some other method. The mode is enabled by the host and hence can be a per pipeline topology attribute. I see no reason not to merge, the conformance test needs fixed to address deep buffer uses cases.

File: include/uapi/sound/sof/abi.h
Comment (lgirdwood) [reply include/uapi/sound/sof/abi.h:32]: yes, looks like  we need a dedicated person to align with me for FW side.

**Review by lgirdwood — APPROVED** [review]

**Comment by yaochunhung** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
@yaochunhung can you rebase as there are conflicts. Thanks !

**Comment by yaochunhung** [comment]
@lgirdwood  I have rebase to resolve the conflicts. Thanks.

**Comment by lgirdwood** [comment]
@plbossart @ranj063 @ujfalusi is there anything blocking this now from kernel side ?

**Review by ranj063 — APPROVED** [review]

**Comment by plbossart** [comment]
> @plbossart @ranj063 @ujfalusi is there anything blocking this now from kernel side ?

there's an unexplained jump from ABI 19 to 21, and no user of that new flag...

**Comment by dbaluta** [comment]
> there's an unexplained jump from ABI 19 to 21, and no user of that new flag...

FW ABI bump that we forgot to expand to Linux kernel ABI is this:

```
commit f6c48532717220a9390e664d6c73e7e6df944206
Author: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>
Date:   Wed Oct 6 10:12:48 2021 -0700

    IPC: DAI: Expand DAI_CONFIG IPC flags
    
    Some DAI components, such as HDaudio, need to be stopped in two steps
    a) stop the DAI component
    b) stop the DAI DMA
    
    This patch enables this two-step stop by expanding the DAI_CONFIG
    IPC flags and split them into 2 parts.
    
    The 4 LSB bits indicate when the DAI_CONFIG IPC is sent, ex: hw_params,
    hw_free or pause. The 4 MSB bits are used as the quirk flags to be used
    along with the command flags. The quirk flag called
    SOF_DAI_CONFIG_FLAGS_2_STEP_STOP shall be set along with the HW_PARAMS
    command flag, i.e. before the pipeline is started so that the stop/pause
    trigger op in the FW can take the appropriate action to either
    perform/skip the DMA stop. If set, the DMA stop will be executed when the
    DAI_CONFIG IPC is sent during hw_free. In the case of pause, DMA pause
    will be handled when the DAI_CONFIG IPC is sent with the PAUSE command
    flag.
    
    Along with this, modify the signature for the hda_ctrl_dai_widget_setup/
    hda_ctrl_dai_widget_free() functions to take additional flags as an
    argument and modify all users to pass the appropriate quirk flags. Only
    the HDA DAI's need to pass the SOF_DAI_CONFIG_FLAGS_2_STEP_STOP quirk
    flag during hw_params to indicate that it supports two-step stop and
    pause.
    
    Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

diff --git a/src/include/kernel/abi.h b/src/include/kernel/abi.h
index b9847c7de..c9db26448 100644
--- a/src/include/kernel/abi.h
+++ b/src/include/kernel/abi.h
@@ -29,7 +29,7 @@
 
 /** \brief SOF ABI version major, minor and patch numbers */
 #define SOF_ABI_MAJOR 3
-#define SOF_ABI_MINOR 19
+#define SOF_ABI_MINOR 20
 #define SOF_ABI_PATCH 0
```

**Comment by lgirdwood** [comment]
> > there's an unexplained jump from ABI 19 to 21, and no user of that new flag...
> 
> FW ABI bump that we forgot to expand to Linux kernel ABI is this:
> 
> ```
> commit f6c48532717220a9390e664d6c73e7e6df944206
> Author: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>
> Date:   Wed Oct 6 10:12:48 2021 -0700
> ```

@ranj063 IIRC this was all part of the DMA flow update for PM (which is now all working), could we be only missing the ABI vesion update here ?

**Comment by dbaluta** [comment]
@yaochunhung can you make a summary of this PR? Is it ready for merge? Are there any open issues?

**Comment by lgirdwood** [comment]
> @yaochunhung can you make a summary of this PR? Is it ready for merge? Are there any open issues?

Ack - lest unblock this ASAP, all ABI is blocked on this.

**Comment by plbossart** [comment]
once https://github.com/thesofproject/linux/pull/3603 is merged, this PR can be rebased and merged.

File: include/uapi/sound/sof/abi.h
Comment (kv2019i) [reply include/uapi/sound/sof/abi.h:32]: @dbaluta @lgirdwood @plbossart This is an open issue. I've made proposal in https://github.com/thesofproject/linux/pull/3298 and https://github.com/thesofproject/linux/pull/3296 , but we didn't reach consensus and no change was done. I believe @ujfalusi has looked at after me. As per current ABI process https://thesofproject.github.io/latest/contribute/process/abiprocess.html , it is not mandatory to update the kernel headers when the FW ABI is extended. It is allowed to delay the update of the kernel headers and do it only when it is needed. This has certain benefits and has been in use for multiple SOF release cycles, but it indeed may lead to conflicts like this one.

**Comment by plbossart** [comment]
@lgirdwood @kv2019i @ujfalusi can you approve #3603 so that this PR can proceed.

**Comment by yaochunhung** [comment]
@lgirdwood @plbossart @dbaluta, I rebase this PR, please help to check it. thanks.

**Comment by miRoox** [comment]
SOFCI TEST

File: include/uapi/sound/sof/abi.h
Comment (plbossart) [inline include/uapi/sound/sof/abi.h:32]: sorry, I screwed up and should have reset this patch level earlier. I added https://github.com/thesofproject/linux/pull/3610 to fix this, you'll have to rebase when it's merged. My bad.

**Review by plbossart — CHANGES_REQUESTED** [review]

File: include/uapi/sound/sof/abi.h
Comment (yaochunhung) [reply include/uapi/sound/sof/abi.h:32]: @plbossart  Never mind, I will rebase this PR when #3610 is merged. Thanks.

Comment (yaochunhung) [reply include/uapi/sound/sof/abi.h:32]: @plbossart  I rebase this PR, thanks.

**Comment by yaochunhung** [comment]
@lgirdwood @plbossart @dbaluta  Please help to check if this PR is good to merge? Thanks.

**Review by dbaluta — APPROVED** [review]

**Review by plbossart — APPROVED** [review]

**Review by ranj063 — APPROVED** [review]

**Comment by dbaluta** [comment]
@lgirdwood there is a boot failure for ADLP not sure it this PR has any thing to do with it.

**Comment by dbaluta** [comment]
@kv2019i ok with you?

**Review by kv2019i — APPROVED** [review]

**Pull request merged**
